### PR TITLE
feat: avoid continuous retries when encountering permanent errors

### PR DESCRIPTION
**Pull request by l-qing**
When the workspace is not configured,
it's a permanent error that cannot be recovered from, so there's no need for repeated retries.

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in

### CODE REVIEW

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 1]
```diff
@@ -755,30 +758,41 @@ func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.Pip
 			continue
 		}
 
+		defer func() {
+			// If it is a permanent error, set the pipelinerun directly to a failed state.
+			if err != nil && controller.IsPermanentError(err) {
+				pr.Status.MarkFailed(ReasonCreateRunFailed, err.Error())
```
Comment (l-qing) [inline pkg/reconciler/pipelinerun/pipelinerun.go:764]: 增加了一个 永久错误的判断，如果是永久错误就直接设置 pipelienrun 的状态为失败。
避免因为 workspace 匹配失败了，还在不停的重试。（这种错误无法恢复）

Comment (l-qing) [reply pkg/reconciler/pipelinerun/pipelinerun.go:764]: 看了下这个代码，没有 RateLimiter，没法获取当前是第几次 reoncile 的地方。
这个代码也没有现成的判断是否是临时错误的函数（可以忽略网络超时导致创建失败等等）

所以在后续中不可恢复的错误包装一下，这里统一处理。

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 2]
```diff
@@ -1050,7 +1064,8 @@ func getTaskrunWorkspaces(ctx context.Context, pr *v1beta1.PipelineRun, rpt *res
 				}
 			}
 			if !workspaceIsOptional {
-				return nil, "", fmt.Errorf("expected workspace %q to be provided by pipelinerun for pipeline task %q", pipelineWorkspace, rpt.PipelineTask.Name)
+				err = fmt.Errorf("expected workspace %q to be provided by pipelinerun for pipeline task %q", pipelineWorkspace, rpt.PipelineTask.Name)
+				return nil, "", controller.NewPermanentError(err)
```
Comment (l-qing) [inline pkg/reconciler/pipelinerun/pipelinerun.go:1068]: 这里产生的错误：`expected workspace "source" to be provided by pipelinerun for pipeline task "xxx"'` 是不可能恢复的，因为 pipelinerun 的 spec 信息也不会变了。

File: pkg/reconciler/pipelinerun/pipelinerun.go (hunk 1 above)
Comment (l-qing) [reply pkg/reconciler/pipelinerun/pipelinerun.go:764]: 类似这种错误在我们那边无法提前判断，因为远程 task 只有在执行时才会获取。

**Review by danielfbm — APPROVED** [review]

**Pull request merged**